lib/spack/spack/environment.py

```diff
@@ -979,8 +996,28 @@ def _read_lockfile_dict(self, d):
                 specs_by_hash[dag_hash]._add_dependency(
                     specs_by_hash[dep_hash], deptypes)
 
-        self.specs_by_hash = dict(
-            (x, y) for x, y in specs_by_hash.items() if x in root_hashes)
+        # If we are reading an older lockfile format (which uses dag hashes
+        # that exclude build deps), we use this to convert the old
+        # concretized_order to the full hashes (preserving the order)
+        old_hash_to_new = {}
+        self.specs_by_hash = {}
+        for _, spec in specs_by_hash.items():
+            old_hash = spec.dag_hash()
+            full_hash = spec.dag_hash(all_deps=True)
```

Comment:
In this case `full_hash` is just the dag hash computed with all dependencies (including build deps). `Spec.full_hash` includes `Package.content_hash`, which is more-specific than we have wanted so far and would be going a bit beyond the original fix.

Given that, I will rename this variable in this context to `build_hash` to avoid confusion.

Reply:
@scheibelp: ok thanks -- that is very clear and makes sense.  Can we:

1. Have 3 methods: `dag_hash()`, `full_hash()`, and `build_hash()`, with comments making them clear, and
2. Implement `full_hash` and `build_hash()` in terms of `dag_hash()`?  I think it would be clearer if `dag_hash()` looked something like this:

    def dag_hash(self, length=none, deptypes=default_deptype, include_package=False):
        deptypes = canonical_deptype(deptypes)
        ...

Then implement `full_hash` and `build_hash` in terms of that.  Does that seem clearer?